File: lab spring25/assignment5/taskD.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
using namespace std;

#define rept(i,n) for(int i=0;i<n;i++)
#define vi vector<int>
#define vb vector<bool>
#define pqi priority_queue<int>
#define br cout << '\n'

void makeGraph(vi *graph,int n, int m){
    pqi g[n+1];
    int u,v;
    rept(i,m){
        cin >> u >> v;
        g[u].push(-v);
    }
    for(int i=1;i<=n; i++){
        pqi &gg = g[i];
        while (!gg.empty()){
            graph[i].push_back(-gg.top());
            gg.pop();
        }
    }
}

void printG(vi *g, int n){
    for(int i=1; i<=n; i++){
        cout << i << ": ";
        for(int u:g[i]) cout << "-> " << u;
        br;
    }br;
}

void bfs(vi *graph,vi &pr, int n, int s, int d){
    if(s==d) return;
    vb vst(n+1,false); vst[s] = true;
    queue<int> Q; Q.push(s);

    while (!Q.empty())
    {
        int at = Q.front(); Q.pop();
        // cout << at << ' ';
        if(at==d) return;
        for(int u:graph[at]) if(!vst[u]) {
            vst[u] = true;
            Q.push(u);
            pr[u] = at;
            if(u==d) return;
        }
    }
}

void path(const vi &pr,stack<int> &pth, int at, int to){
    while (at!=to && at!=0){
        pth.push(at);
        at = pr[at];
    }
}

void solve(){
    int n,m,s,d,k;
    cin >> n >> m >> s >> d >> k;
    vi graph[n+1];
    makeGraph(graph,n,m);
    // printG(graph,n);
    
    if(s==k && k==d){
        cout << "0\n" << s << '\n';
        return;
    }
    vi pr(n+1);
    stack<int> pth;
    if(k!=d){
        bfs(graph,pr,n,k,d);
        if(pr[d]==0){
            cout << "-1\n";
            return;
        }
        path(pr,pth,d,k);
    }
    rept(i,n+1) pr[i]=0;
    
    if(s!=k){
        bfs(graph,pr,n,s,k);
        if(pr[k]==0){
            cout << "-1\n";
            return;
        }
        path(pr,pth,k, s);
    }

    pth.push(s);
    cout << pth.size()-1 << '\n';
    while(!pth.empty()) {
        cout << pth.top() << ' ';
        pth.pop();
    }br;
}

int main(){
    ios::sync_with_stdio(false);cin.tie(nullptr);cout.tie(nullptr);
    freopen("input.txt","r",stdin);
    int t{1};
    cin >> t;
    while(t--) solve();
}

File: lab summer25/assignment5/taskD.py
import sys
from collections import deque

sys.stdin = open("input.txt","r")

def bfs(graph,n,s,d):
    if s==d:
        path = deque()
        path.append(s)
        return 0,path

    q = deque()
    visited = [False]*(n+1)
    q.append(s)
    visited[s] = True
    parent = [0]*(n+1)
    found = False
    while q:
        at = q.popleft()
        for v in graph[at]:
            if visited[v]: continue
            visited[v] = True
            q.append(v)
            parent[v] = at
            found = (v==d)
            if found: break
        if found: break
    
    if not found: return (-1,None)
    path = deque()
    at = d
    while parent[at]!=0:
        path.append(at)
        at = parent[at]
    path.append(s)
    return len(path)-1,path

def solve():
    n,m,s,d,k = list(map(int,input().split()))
    graph = [[] for _ in range(n+1)]
    for _ in range(m):
        u,v = list(map(int,input().split()))
        graph[u].append(v)
    
    len_sk,sk = bfs(graph,n,s,k)
    len_kd,kd = bfs(graph,n,k,d)
    if len_sk<0 or len_kd<0:
        print('-1')
        return
    print(len_sk+len_kd)
    for _ in range(len_sk): print(sk.pop(), end=" ")
    while kd: print(kd.pop(), end=" ")

t = 1
t = int(input())
while t>0:
    t-=1
    solve()
    print()